clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp

```diff
@@ -966,9 +1035,18 @@ class BinaryWrapper {
         // Adding ELF notes for STDIN is not supported yet.
         Bin = addELFNotes(Bin, Img.File);
       }
-      std::pair<Constant *, Constant *> Fbin = addDeviceImageToModule(
-          ArrayRef<char>(Bin->getBufferStart(), Bin->getBufferSize()),
-          Twine(OffloadKindTag) + Twine(ImgId) + Twine(".data"), Kind, Img.Tgt);
+      std::pair<Constant *, Constant *> Fbin;
+      if (NativeCPU) {
```

Comment:
Does this way of integrating `native_cpu` target allows for a single call to `clang-offload-wrapper` to wrap all targets at the same time?

I haven't reviewed the driver changes, but based on this, I assume that we invoking `clang-offload-wrapper` separately for `native_cpu`, right?

I wonder, can this `if` be implemented based on a target for the currently processed image?

Reply:
> Does this way of integrating native_cpu target allows for a single call to clang-offload-wrapper to wrap all targets at the same time?

This should be possible but we haven't looked into that yet. But we'll put that on the todo list and create another PR with dedicated clang-offload-wrapper tests if that's ok?

> I haven't reviewed the driver changes, but based on this, I assume that we invoking clang-offload-wrapper separately for native_cpu, right?

Yes, that's right. However, it seems that for a normal clang invocation requesting multiple targets (even without native_cpu) the clang-offload-wrapper is currently invoked for each target individually. Or is there another option to generate only one call to `clang-offload-wrapper`?  

> I wonder, can this if be implemented based on a target for the currently processed image?

Great find! The option is now removed and the check is based on the image as you suggested. Thank You!

Reply:
> Yes, that's right. However, it seems that for a normal clang invocation requesting multiple targets (even without native_cpu) the `clang-offload-wrapper` is currently invoked for each target individually. Or is there another option to generate only one call to `clang-offload-wrapper`?

I honestly don't remember that implementation detail of the toolchain, better ask @mdtoguchi here. If we already launch `clang-offload-wrapper` individually for each target, then I don't think that we should do something special here for `native_cpu`.

My main point was to ensure that all targets are handled in a similar manner, that's all.

> But we'll put that on the todo list and create another PR with dedicated `clang-offload-wrapper` tests if that's ok?

As I've said, it is not even needed if we don't do  that for other targets already. I mean, for the purpose of optimizing the toolchain we may consider doing so, but I believe that it will be done automatically once we switch to `clang-linker-wrapper`. So, at the moment it is more of: we need to keep that in mind when further working on `native_cpu` target

Reply:
The wrapper is called for each target individually, so this is fine.  There is a way to do a singular call for multiple targets, but the current toolchain is not setup to take advantage of this at the moment.

Reply:
Thank you very much @AlexeySachkov for raising this and @mdtoguchi for the info, we'll keep an eye on this. If `native_cpu` causes any issues with `clang-offload-wrapper` or `clang-linker-wrapper`  please let us know. Thank you!